### ISSUE TITLE
Blocks don't refresh after add

### ISSUE THREAD

**Issue by ScopeyNZ** [dnadesign/silverstripe-elemental#312]
On **master**.

After adding a block using the add tool - when you click the breadcrumb to return to the previous page the new block does not show - but it will after a reload.

**Comment by robbieaverill** [comment]
This is a common problem with the way we shiv in React components in an entwine based environment. Another example of this is the history viewer. If you use it, go back to Content tab then save a couple of times and go back the new versions aren't shown yet.

The content that the ReactDOM shiv matches is loaded when the form is loaded so the editor is instantiated early on and not always unmounted. I'm not sure what a good fix for this is yet

**Comment by NightJar** [comment]
Steps to reproduce:

```
    "require": {
        "php": ">=5.6.0",
        "silverstripe/recipe-plugin": "^1.2",
        "silverstripe/recipe-cms": "4.3.x-dev",
        "silverstripe-themes/simple": "~3.2.0",
        "silverstripe/recipe-content-blocks": "2.x-dev",
        "dnadesign/silverstripe-elemental": "4.x-dev as 3.x-dev"
    },
```

👇 Do
- **Load** `site/dev/build` && **Load** `site/admin/pages`
- **Click** `Add new` _Blocks Page_
- **Select** `Content` from `(Select type to create)` && **Click** `Add`
- **Click** `Create`
- **Click** either the back caret `<` _or_ `New Blocks Page` breadcrumb from the `cms-content-header north`

👀 No blocks are listed.

**Comment by robbieaverill** [comment]
So there are a few places where this flow breaks:

* Adding a new block via PHP GridField (current way to do it - new block doesn't show up)
* Publishing the page (block states don't update until refresh)
* Adding a new block then going to the history tab (new versions don't show up until refresh)

**Comment by brynwhyman** [comment]
#340 could fix this

**Comment by raissanorth** [comment]
Had a look at this using #340.

Blocks now refresh after 

- saving an individual block (inline)
- saving the page
- publishing the page

Blocks however still don't refresh after publishing individual blocks (inline)... 😞

**Comment by NightJar** [comment]
Publishing.

Publishing is hard.

Publishing launches an graphql mutation to copy a particular version from stage to live.
The problem with this is that it does not factor in any 'dirty' state of the form for that element when inline editing.
That makes 1 requests.

We cannot submit data for saving through a mutation because each block is different and graphql queries are strongly defined & typed.
So we must call a REST endpoint to perform the save.
That makes 2 requests.

The publish action is built with the version number of the element it is related to. This then needs updating, which means refreshing the apollo store as the save action does.
That makes 3 requests.

The mutation for publishing then automatically refetches the data for the page through another query.
That makes 4 requests.

Once that's all done, there's another request to update the preview pane.



This cannot be sane.

**Comment by NightJar** [comment]
# TODO

- [ ] Give the publish action some way of detecting a 'dirty' form
- [x] Make publishes save before publishing, iff the form is 'dirty'

**Comment by NightJar** [comment]
OK, so all that works, all on #467 
But it works a little too well in that form updates are also tracked by redux form if the _structure_ changes
e.g. someone shows a form - it loads into state.
Someone then hides that form - all the `registeredFields` go from `count: 1` to `count: 0`, which then triggers the publish action to show (and save actions to fire - doubling the request count for a publish action)

**Comment by NightJar** [comment]
Chasing this last improvement is sinking time.
The `updatedForms` has the form set in it.
The form state has initial with the original value.
The form state has values with the _correct_ updated state.
Yet passing this state into a `isDirty(formName)(state)` results in `false`. Most bizarre.
I don't have time to debug a third party library. It works for now, and we can make an issue to improve this in future.

**Comment by NightJar** [comment]
#470 tracks the remaining performance issues.
I think #467 should be able to be merged, baring review feedback.

**Comment by ScopeyNZ** [comment]
Closed in #467
